distributed/shuffle/tests/test_shuffle.py

```diff
@@ -469,3 +475,279 @@ async def test_clean_after_close(c, s, a, b):
 
     # clean_scheduler(s)  # TODO
     # clean_worker(b)  # TODO
+
+
+class PooledRPCShuffle(PooledRPCCall):
+    def __init__(self, shuffle: Shuffle):
+        self.shuffle = shuffle
+
+    def __getattr__(self, key):
+        async def _(**kwargs):
+            from distributed.protocol.serialize import nested_deserialize
+
+            method_name = key.replace("shuffle_", "")
+            kwargs.pop("shuffle_id", None)
+            # TODO: This is a bit awkward. At some point the arguments are
+            # already getting wrapped with a `Serialize`. We only want to unwrap
+            # here.
+            kwargs = nested_deserialize(kwargs)
+            meth = getattr(self.shuffle, method_name)
+            return await meth(**kwargs)
+
+        return _
+
+
+class ShuffleTestPool:
+    def __init__(self, *args, **kwargs):
+        self.shuffles = {}
+        super().__init__(*args, **kwargs)
+
+    def __call__(self, addr: str, *args: Any, **kwargs: Any) -> PooledRPCShuffle:
+        return PooledRPCShuffle(self.shuffles[addr])
+
+    async def fake_broadcast(self, msg):
+
+        op = msg.pop("op").replace("shuffle_", "")
+        out = {}
+        for addr, s in self.shuffles.items():
+            out[addr] = await getattr(s, op)()
+        return out
+
+    def new_shuffle(
+        self, name, worker_for_mapping, schema, directory, loop, Shuffle=Shuffle
+    ):
+        s = Shuffle(
+            column="_partition",
+            worker_for=worker_for_mapping,
+            # FIXME: Is output_workers redundant with worker_for?
+            output_workers=set(worker_for_mapping.values()),
+            schema=schema,
+            directory=directory / name,
+            id=ShuffleId(name),
+            local_address=name,
+            nthreads=2,
+            rpc=self,
+            loop=loop,
+            broadcast=self.fake_broadcast,
+        )
+        self.shuffles[name] = s
+        return s
```

Comment:
This might be the most interesting part of this refactoring. I separated `Shuffle` and `ShuffleWorkerExtension` sufficiently to allow us to write down tests without scheduler and workers. This allows for a much more granular control over concurrency issues.

Particularly the offload tests below produce race conditions that would drop data